scipy/spatial/transform/_rotation.pyx

```diff
@@ -2580,31 +2580,39 @@ cdef class Rotation:
             raise ValueError("Expected input of shape (3,) or (P, 3), "
                              "got {}.".format(vectors.shape))
 
+        n_vectors = vectors.shape[0]
```

Comment:
I would put

```
single_vector = vectors.ndim == 1
n_vectors = 1 if single_vector else len(vectors)
```

Also some reasons to use types for all variables, can matter if `apply` is called often I suppose.

Reply:
I can add those changes.

I am not entirely sure what you mean by types for all variables.

However, trying to understand what you mean I actually realized that if you apply a rotation to a `np.float32` vector, you will get a `float` vector back. This is how it has been done so far, but it could be unanticipated and lead to memory issues for very large vectors. If you like me to change that, let me know, both `np.matmul` and `np.einsum` take a `dtype` kwarg, however it would affect the precision of the result, as both the rotation matrix and the vector are "cast" to the new dtype before performing the operation (at least that is my limited understanding of what is happening).

Reply:
I meant a small thing, that `n_vectors`, `single_vector` probably should be defined with `cdef Py_ssize_t`, etc..

But as there are many small non-optimal things involving Python interaction it doesn't look necessary.

Reply:
I see, I didn't even consider cython types because there were none before. I added them, let me know if there is anything wrong with them, I am a absolute cython beginner.

Reply:
I implemented the types and the requested changes. Let me know if there is anything else that I should work on.